internal/controller/controller_functions_test.go

```diff
@@ -0,0 +1,176 @@
+
+func (m *MockStatusWriter) Update(ctx context.Context, obj client.Object, opts ...client.SubResourceUpdateOption) error {
+	args := m.Called(ctx, obj)
+
+	return args.Error(0)
+}
+
+func (m *MockStatusWriter) Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error {
+	args := m.Called(ctx, obj, patch)
+
+	return args.Error(0)
+}
+
+// MockClient is a mock implementation of the client.Client interface for testing
+type MockClient struct {
+	mock.Mock
+	client.Client
+	statusWriter client.StatusWriter
+}
+
+func (m *MockClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	args := m.Called(ctx, key, obj)
+	if dep, ok := obj.(*appsv1.Deployment); ok {
+		*dep = appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{Name: key.Name, Namespace: key.Namespace},
+			Spec: appsv1.DeploymentSpec{
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{Annotations: make(map[string]string)},
+				},
+			},
+		}
+	} else if rule, ok := obj.(*karov1alpha1.RestartRule); ok {
+		*rule = karov1alpha1.RestartRule{ObjectMeta: metav1.ObjectMeta{Name: key.Name, Namespace: key.Namespace}}
+	}
+
+	return args.Error(0)
+}
+
+func (m *MockClient) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	args := m.Called(ctx, obj)
+
+	return args.Error(0)
+}
+
+//nolint:ireturn
+func (m *MockClient) Status() client.StatusWriter {
+	if m.statusWriter == nil {
+		m.statusWriter = &MockStatusWriter{}
+	}
+
+	return m.statusWriter
+}
+
+func TestProcessRestartRules(t *testing.T) {
+	t.Parallel()
+
+	rule1 := &karov1alpha1.RestartRule{
+		ObjectMeta: metav1.ObjectMeta{Name: "rule1", Namespace: "default"},
+		Spec:       karov1alpha1.RestartRuleSpec{Targets: []karov1alpha1.TargetSpec{{Kind: "Deployment", Name: "test-deployment"}}},
+	}
+	rule2 := &karov1alpha1.RestartRule{
+		ObjectMeta: metav1.ObjectMeta{Name: "rule2", Namespace: "default"},
+		Spec:       karov1alpha1.RestartRuleSpec{Targets: []karov1alpha1.TargetSpec{{Kind: "Deployment", Name: "test-deployment"}}},
+	}
+
+	tests := []struct {
+		name                 string
+		restartRules         []*karov1alpha1.RestartRule
+		setupMocks           func(*MockClient, *MockStatusWriter)
+		expectedDeployment   int
+		expectedStatusUpdate int
+		expectedErr          bool
+	}{
+		{
+			name:                 "prevent duplicate restarts",
+			restartRules:         []*karov1alpha1.RestartRule{rule1, rule2},
+			expectedDeployment:   1,
+			expectedStatusUpdate: 2,
+			expectedErr:          false,
+			setupMocks: func(mockClient *MockClient, mockStatusWriter *MockStatusWriter) {
+				mockClient.On("Get", mock.Anything, mock.Anything, mock.Anything).Return(nil)
+				mockClient.On("Update", mock.Anything, mock.AnythingOfType("*v1.Deployment")).Return(nil)
+				mockStatusWriter.On("Update", mock.Anything, mock.AnythingOfType("*v1alpha1.RestartRule")).Return(nil)
+			},
+		},
+		{
```

Comment:
Please add a test case for the plain vanilla case: Two rules with different targets, no duplicate restart expected